onnxruntime/core/optimizer/identical_children_consolidation.cc

```diff
@@ -0,0 +1,128 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "core/optimizer/identical_children_consolidation.h"
+
+#include "core/graph/graph_utils.h"
+#include "core/optimizer/initializer.h"
+#include "core/optimizer/utils.h"
+
+
+namespace onnxruntime {
+Status IdenticalChildrenConsolidation::ApplyImpl(Graph& graph, bool& modified, int /*graph_level*/, const logging::Logger& /*logger*/) const {
+  GraphViewer const graph_viewer(graph);
+  for (auto node_index : graph_viewer.GetNodesInTopologicalOrder()) {
+    Node* node = graph.GetNode(node_index);
+    if (!IsSupportedParentNode(node)) {
+      continue;
+    }
+    for (auto supported_op : supported_ops.at(node->OpType())) {
+      for (auto twin_group : DivideIdenticalChildrenIntoGroups(graph, node, supported_op)) {
+        // If there is no twins in the group, skip it.
+        if (twin_group.size() <= 1) {
+          continue;
+        }
+        Node* first_twin = graph.GetNode(twin_group[0]);
+        for (size_t i = 1; i < twin_group.size(); i++) {
+          Node* other_twin = graph.GetNode(twin_group[i]);
+          // Replace all outputs of other_twin with first_twin
+          // If A = parent_node, B1, B2 are twins, C1, C2 are grandchild from each twin
+          // We want 2 edges B1 -> C1 and  B2->C2 to become => B1->C1, B1->C2
+          // So ReplaceDownstreamNodeInput(graph, B2,j, B1,j) is called
+          for (size_t j = 0; j < other_twin->GetOutputEdgesCount(); j++) {
+            graph_utils::ReplaceDownstreamNodeInput(graph, *other_twin, static_cast<int>(j), *first_twin, static_cast<int>(j));
+          }
```

Comment:
This is incorrect. ReplaceDownstreamNodeInput takes in arg index instead of the index of edge. 

For the current case you handle (Dequantize and QuantizeLinear), the index is 0 and you don't need a loop.

Reply:
In addition, you need to check and make sure other_twin's output is not used as graph output.